### PR TITLE
Decouple Google_Client creation from OAuth_Client class

### PR DESCRIPTION
**Pull request by felixarntz**
## Summary

Addresses issue #3658

## Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 4.7 and PHP 5.6.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

### CODE REVIEW

File: tests/phpunit/includes/Core/Authentication/Clients/OAuth_Client_Empty.php [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+ * @copyright 2021 Google LLC
+ * @license   https://www.apache.org/licenses/LICENSE-2.0 Apache License 2.0
+ * @link      https://sitekit.withgoogle.com
+ */
+
+namespace Google\Site_Kit\Tests\Core\Authentication\Clients;
+
+use Google\Site_Kit\Core\Authentication\Clients\OAuth_Client_Base;
+
+class OAuth_Client_Empty extends OAuth_Client_Base {
```
Comment (aaemnnosttv) [inline tests/phpunit/includes/Core/Authentication/Clients/OAuth_Client_Empty.php:15]: Why not use a mock instance instead? It essentially does the same thing just on-the-fly and gives us more control without the need for an added class just for tests.

File: tests/phpunit/integration/Core/Authentication/TokenTest.php [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+		$token = new Token( $this->user_options );
+
+		// Cannot set empty token data.
+		$this->assertFalse( $token->set( array() ) );
+
+		// Cannot set token data without access token.
+		$this->assertFalse( $token->set( array( 'refresh_token' => 'h8i9j0' ) ) );
+		$this->assertFalse( $token->set( array( 'expires_in' => 3600 ) ) );
+
+		// Setting only access token also sets sane defaults for created and expires_in.
```
Comment (aaemnnosttv) [inline tests/phpunit/integration/Core/Authentication/TokenTest.php:102]:
```suggestion
		// Setting only access token also sets correct defaults for created and expires_in.
```

**Review by aaemnnosttv — CHANGES_REQUESTED** [review]
Thanks @felixarntz ! Overall this looks really good. I've left a few comments, questions, and suggestions for you below but very little of it is blocking.

File: includes/Core/Authentication/Clients/OAuth_Client.php [hunk 3]
```diff
@@ -17,115 +17,27 @@
 use Google\Site_Kit\Core\Authentication\Google_Proxy;
 use Google\Site_Kit\Core\Authentication\Owner_ID;
 use Google\Site_Kit\Core\Authentication\Profile;
+use Google\Site_Kit\Core\Authentication\Token;
 use Google\Site_Kit\Core\Permissions\Permissions;
 use Google\Site_Kit\Core\Storage\Encrypted_Options;
```
[the remark this answers is not in the document]
Comment (felixarntz) [reply includes/Core/Authentication/Clients/OAuth_Client.php:22]: It is I think, in the constructor and its docs 🤔

File: includes/Core/Authentication/Clients/OAuth_Client.php [hunk 4]
```diff
@@ -444,7 +174,7 @@ public function get_granted_additional_scopes() {
 	 * @return bool true if any required scopes are not satisfied, otherwise false.
 	 */
 	public function needs_reauthentication() {
-		if ( ! $this->get_access_token() ) {
+		if ( ! $this->get_token() ) {
```
[the remark this answers is not in the document]
Comment (felixarntz) [reply includes/Core/Authentication/Clients/OAuth_Client.php:177]: I think it makes sense to use `$this->token->has()` instead, which is now the most appropriate way to see if the user is authenticated, which technically is what we want to know here. See also the change to `Authentication::is_authenticated`.

File: includes/Core/Authentication/Clients/OAuth_Client_Base.php [hunk 5]
```diff
@@ -0,0 +1,348 @@
...
+ * @since n.e.x.t
+ * @access private
+ * @ignore
+ */
+abstract class OAuth_Client_Base {
+
+	const OPTION_ACCESS_TOKEN            = 'googlesitekit_access_token';
+	const OPTION_ACCESS_TOKEN_EXPIRES_IN = 'googlesitekit_access_token_expires_in';
+	const OPTION_ACCESS_TOKEN_CREATED    = 'googlesitekit_access_token_created_at';
+	const OPTION_REFRESH_TOKEN           = 'googlesitekit_refresh_token';
```
[the remark this answers is not in the document]
Comment (felixarntz) [reply includes/Core/Authentication/Clients/OAuth_Client_Base.php:37]: Yeah I left them there for BC. Not super critical, but I also don't think we really should move them to the new `Token` class since we eventually want to consolidate the options anyway. I'd prefer to therefore just leave them as is at this point. The BC considerations are small, but also moving them wouldn't be that beneficial.

File: includes/Core/Authentication/Token.php [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+	private $encrypted_user_options;
+
+	/**
+	 * User_Setting constructor.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param User_Options_Interface $user_options User_Options_Interface instance.
+	 */
+	public function __construct( User_Options_Interface $user_options ) {
```
[the remark this answers is not in the document]
Comment (felixarntz) [reply includes/Core/Authentication/Token.php:53]: Good point - however, why is that though? Is there a good reason or was that just an oversight? Looking at it now, I think it might be most appropriate to change `Encrypted_User_Options` to expect `User_Options_Interface` instead?

File: includes/Core/Authentication/Token.php [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+		$token = array(
+			'access_token' => $access_token,
+			'expires_in'   => (int) $this->user_options->get( OAuth_Client::OPTION_ACCESS_TOKEN_EXPIRES_IN ),
+			'created'      => (int) $this->user_options->get( OAuth_Client::OPTION_ACCESS_TOKEN_CREATED ),
+		);
+
+		$refresh_token = $this->encrypted_user_options->get( OAuth_Client::OPTION_REFRESH_TOKEN );
+		if ( ! empty( $refresh_token ) ) {
+			$token['refresh_token'] = $refresh_token;
+		}
```
[the remark this answers is not in the document]
Comment (felixarntz) [reply includes/Core/Authentication/Token.php:95]: This implementation is in line with how the OAuth client works. The token data only includes a `refresh_token` key if there is one, so I think it makes sense to mirror that. Also, without an `access_token` the entire data is invalid, so I don't think we should in that case return that state.

File: includes/Core/Authentication/Token.php [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+
+		return $token;
```
[the remark this answers is not in the document]
Comment (felixarntz) [reply includes/Core/Authentication/Token.php:97]: See above.

File: includes/Core/Authentication/Token.php [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+
+	/**
+	 * Checks whether or not the setting exists.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return bool True on success, false on failure.
+	 */
+	public function has() {
+		if ( ! $this->get() ) {
```
[the remark this answers is not in the document]
Comment (felixarntz) [reply includes/Core/Authentication/Token.php:66]: See above.

File: includes/Core/Authentication/Token.php [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+	 * @param mixed $value Setting value. Must be serializable if non-scalar.
+	 *
+	 * @return bool True on success, false on failure.
+	 */
+	public function set( $value ) {
+		if ( ! is_array( $value ) || empty( $value['access_token'] ) ) {
+			return false;
+		}
+
+		// Use sane defaults for these fields.
```
[the remark this answers is not in the document]
Comment (felixarntz) [reply includes/Core/Authentication/Token.php:114]: Good catch! 👍

File: includes/Core/Authentication/Token.php [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+			$value['created'] = time();
+		}
+
+		$this->encrypted_user_options->set( OAuth_Client::OPTION_ACCESS_TOKEN, $value['access_token'] );
+		$this->user_options->set( OAuth_Client::OPTION_ACCESS_TOKEN_EXPIRES_IN, $value['expires_in'] );
+		$this->user_options->set( OAuth_Client::OPTION_ACCESS_TOKEN_CREATED, $value['created'] );
+
+		if ( ! empty( $value['refresh_token'] ) ) {
+			$this->encrypted_user_options->set( OAuth_Client::OPTION_REFRESH_TOKEN, $value['refresh_token'] );
+		}
```
[the remark this answers is not in the document]
Comment (felixarntz) [reply includes/Core/Authentication/Token.php:128]: Yes, what is there now is the expected behavior according to how the OAuth client works. An OAuth API response itself even _may_ not include the refresh token in a request to refresh an access token. Surely we could enforce the existing refresh token to be added to the returned data, but the common behavior is to maintain the existing refresh token in the data, unless it explicitly needs to be overridden.

In other words, yes a refresh token only gets returned by Google's OAuth infrastructure when going through the OAuth authorization flow.

File: tests/phpunit/includes/Core/Authentication/Clients/OAuth_Client_Empty.php (hunk 1 above)
Comment (felixarntz) [reply tests/phpunit/includes/Core/Authentication/Clients/OAuth_Client_Empty.php:15]: Hmm, I may not be aware of that possibility. Do we do that anywhere yet? How can I create a mock instance of an abstract class?

**Comment by felixarntz** [comment]
@aaemnnosttv I've addressed most of your feedback and left a few responses on the other feedback. This should be ready for another pass.

File: includes/Core/Authentication/Clients/OAuth_Client.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (aaemnnosttv) [reply includes/Core/Authentication/Clients/OAuth_Client.php:22]: It's the only usage in the file 🙂  Did you maybe think I was referring to `Token`?
```
⚡  grep Encrypted_Options includes/Core/Authentication/Clients/OAuth_Client.php                 
use Google\Site_Kit\Core\Storage\Encrypted_Options;
```

File: includes/Core/Authentication/Token.php (hunk 6 above)
[the remark this answers is not in the document]
Comment (aaemnnosttv) [reply includes/Core/Authentication/Token.php:53]: `User_Options` and `Encrypted_User_Options` both implement the `User_Options_Interface`. So if we changed `Encrypted_User_Options` to accept `User_Options_Interface` instead, then it could technically be instantiated with an instance of `Encrypted_User_Options` which of course wouldn't work, hence breaking the contract.

So in this case, the `User_Options_Interface` abstracts the storage implementation which includes (un)encrypted data, but `Encrypted_User_Options` needs to read and write "raw" values. As lower-level classes, I think this is okay given our current architecture.

To your point, I think we would need new interfaces for plaintext and encrypted storage in order to allow these dependencies to be passed fulfilling an interface rather than a concrete implementation.

File: includes/Core/Authentication/Token.php (hunk 7 above)
[the remark this answers is not in the document]
Comment (aaemnnosttv) [reply includes/Core/Authentication/Token.php:95]: Ah, you're referring to `Google\Client` correct? If so, I think I see what you mean.

We could use `null` for the empty value here which would be interpreted the same within the `Client` because it uses `isset` everywhere.

If not, then this is fine 👍

File: tests/phpunit/includes/Core/Authentication/Clients/OAuth_Client_Empty.php (hunk 1 above)
Comment (aaemnnosttv) [reply tests/phpunit/includes/Core/Authentication/Clients/OAuth_Client_Empty.php:15]: Ah, it would be the same as creating a mock for any other class. PHPUnit creates a mock by creating a new class that extends the class it mocks so nothing special would be needed for an abstract class, so long as it didn't also have abstract methods – in which case I think the mock would just need to have those defined in the definition.

So I think we'd just need to `$this->getMockBuilder( OAuth_Client_Base::class )->getMock()`.  There's also a `createMock` method but it does a but more like disabling the constructor.

File: includes/Core/Authentication/Token.php (hunk 6 above)
[the remark this answers is not in the document]
Comment (aaemnnosttv) [reply includes/Core/Authentication/Token.php:53]: I think this is the only thing left to address.

**Review by aaemnnosttv — COMMENTED** [review]
Thanks @felixarntz – this is super close. I left replies to a few comments from the last round with just one thing left that should probably change.

File: includes/Core/Authentication/Clients/OAuth_Client.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (felixarntz) [reply includes/Core/Authentication/Clients/OAuth_Client.php:22]: Ahh, yes I indeed looked at the wrong thing.

File: includes/Core/Authentication/Token.php (hunk 7 above)
[the remark this answers is not in the document]
Comment (felixarntz) [reply includes/Core/Authentication/Token.php:95]: I think it makes sense to keep as is to follow the client behavior - technically `null` would work, but I guess with `array_key_exists` being a thing it is still technically a slightly different format.

File: includes/Core/Authentication/Token.php [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+ * so that the four pieces of data become a single user setting.
+ *
+ * @since n.e.x.t
+ * @access private
+ * @ignore
+ */
+final class Token {
+
+	/**
+	 * User_Options implementation.
```
Comment (aaemnnosttv) [inline includes/Core/Authentication/Token.php:31]:
```suggestion
	 * User_Options instance.
```

**Review by aaemnnosttv — APPROVED** [review]
Great work @felixarntz !
